File: backend/app/self_dimension_logic.py
"""套一六维「底层逻辑」Layer B：分数带 × 维度词库 → 结构化解读。"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

LOGIC_LIBRARY_PATH = Path(__file__).resolve().parents[1] / "data" / "self_dimension_logic_v1.json"

POSITION_LABELS: dict[str, tuple[str, str]] = {
    "SA2": ("高焦虑", "低焦虑"),
    "SA3": ("高回避", "低回避"),
}


@lru_cache(maxsize=1)
def load_dimension_logic_library() -> dict[str, Any]:
    try:
        return json.loads(LOGIC_LIBRARY_PATH.read_text(encoding="utf-8"))
    except OSError:
        return {"dimensions": {}}


def score_band_key(score: float) -> str:
    s = max(0, min(100, round(float(score))))
    if s <= 30:
        return "emerging"
    if s <= 50:
        return "developing"
    if s <= 65:
        return "steady"
    if s <= 80:
        return "asset"
    return "strength"


def position_labels(code: str) -> tuple[str, str]:
    return POSITION_LABELS.get(code, ("还在展开", "更成熟"))


def build_underlying_logic(code: str, score: float) -> dict[str, Any] | None:
    lib = load_dimension_logic_library()
    dim = (lib.get("dimensions") or {}).get(code)
    if not dim:
        return None
    band_key = score_band_key(score)
    band = (dim.get("bands") or {}).get(band_key)
    if not band:
        return None
    return {
        "measure": str(dim.get("measure") or ""),
        "headline": str(band.get("headline") or ""),
        "interpretation": str(band.get("interpretation") or ""),
        "inRelationship": str(band.get("inRelationship") or ""),
        "growthHint": band.get("growthHint"),
        "bandKey": band_key,
    }


def enrich_dimension_summary(
    code: str,
    score: float,
    meta: dict[str, str],
    *,
    label: str,
) -> dict[str, Any]:
    """Merge scoring metadata with Layer B underlying-logic copy."""
    raw = max(0, min(100, round(float(score))))
    low, high = position_labels(code)
    logic = build_underlying_logic(code, raw)
    headline = (logic or {}).get("headline") or ""
    return {
        "label": label,
        "detail": headline,
        "position": raw,
        "position_label_low": low,
        "position_label_high": high,
        "name": meta.get("name") or code,
        "coreQuestion": meta.get("core") or "",
        "underlying_logic": logic,
    }

File: frontend/src/lib/selfDimensionLogic.ts
import logicLibrary from "@/data/self_dimension_logic_v1.json";
import type { SelfDimensionCode } from "@/data/selfSuiteSpec";

export type SelfUnderlyingLogic = {
  measure: string;
  headline: string;
  interpretation: string;
  inRelationship: string;
  growthHint?: string | null;
  bandKey: string;
};

type BandKey = "emerging" | "developing" | "steady" | "asset" | "strength";

function scoreBandKey(score: number): BandKey {
  const s = Math.max(0, Math.min(100, Math.round(score)));
  if (s <= 30) return "emerging";
  if (s <= 50) return "developing";
  if (s <= 65) return "steady";
  if (s <= 80) return "asset";
  return "strength";
}

/** 与 backend/app/self_dimension_logic.py 同词库、同分数带规则。 */
export function resolveSelfUnderlyingLogic(
  code: SelfDimensionCode,
  score: number,
): SelfUnderlyingLogic | null {
  const dim = logicLibrary.dimensions[code as keyof typeof logicLibrary.dimensions];
  if (!dim) return null;
  const bandKey = scoreBandKey(score);
  const band = dim.bands[bandKey as keyof typeof dim.bands];
  if (!band) return null;
  return {
    measure: dim.measure,
    headline: band.headline,
    interpretation: band.interpretation,
    inRelationship: band.inRelationship,
    growthHint: band.growthHint ?? null,
    bandKey,
  };
}

export function positionLabelsForDimension(code: string): { low: string; high: string } {
  if (code === "SA2") return { low: "高焦虑", high: "低焦虑" };
  if (code === "SA3") return { low: "高回避", high: "低回避" };
  const spec = logicLibrary.dimensions[code as SelfDimensionCode];
  if (!spec) return { low: "偏低", high: "偏高" };
  return { low: "还在展开", high: "更成熟" };
}
